File: frontend/src/objects/contest/ContestSlug.ts
/** 比赛公开 slug 品牌类型；用于 URL 和公开 API 路径。 */
export type ContestSlug = string & { readonly __brand: 'ContestSlug' }

/** 解析 helper 的返回结果；成功携带品牌值，失败携带可展示错误，不抛异常。 */
type ParseResult<T> = { ok: true; value: T } | { ok: false; error: string }

const slugPattern = /^[a-z0-9]+(?:-[a-z0-9]+)*$/

/** 创建比赛 slug 品牌值；调用前必须完成格式和长度校验。 */
function createContestSlug(value: string): ContestSlug {
  /** 注意：这里的 as 只在 parseContestSlug 校验通过后施加品牌类型。 */
  return value as ContestSlug
}

/** 将比赛 slug 还原为字符串；slug 语法已限制为 URL 安全字符，调用方可按路径或 query 场景继续编码。 */
export function contestSlugValue(slug: ContestSlug): string {
  return slug
}

/** 解析比赛 slug；拒绝空值、过短/过长值和非法字符。 */
export function parseContestSlug(rawSlug: string): ParseResult<ContestSlug> {
  const normalized = rawSlug.trim()
  if (!normalized) {
    return { ok: false, error: 'Contest slug is required.' }
  }
  if (normalized.length < 3 || normalized.length > 64) {
    return { ok: false, error: 'Contest slug must be between 3 and 64 characters.' }
  }
  if (!slugPattern.test(normalized)) {
    return { ok: false, error: 'Contest slug may contain only lowercase letters, numbers, and hyphens.' }
  }
  return { ok: true, value: createContestSlug(normalized) }
}

File: backend/src/main/scala/domains/contest/objects/ContestSlug.scala
package domains.contest.objects

import io.circe.{Decoder, Encoder}

/** 比赛 URL slug 领域值，用于公开路径和内部查找。 */
final case class ContestSlug(value: String)

/** 提供比赛 slug JSON codec 和路径安全格式校验。 */
object ContestSlug:
  given Encoder[ContestSlug] = Encoder.encodeString.contramap(_.value)
  given Decoder[ContestSlug] = Decoder.decodeString.emap(parse)

  private val slugPattern = "^[a-z0-9]+(?:-[a-z0-9]+)*$".r

  /** 去除首尾空白，要求 3-64 位小写字母、数字和中横线组合。 */
  def parse(raw: String): Either[String, ContestSlug] =
    val normalized = raw.trim
    if normalized.isEmpty then Left("Contest slug is required.")
    else if normalized.length < 3 || normalized.length > 64 then Left("Contest slug must be between 3 and 64 characters.")
    else if !slugPattern.matches(normalized) then Left("Contest slug may contain only lowercase letters, numbers, and hyphens.")
    else Right(ContestSlug(normalized))
